src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,387 @@
+
+    private boolean dominates(Pointer source, Pointer target) {
+        int dominates = 0;
+
+        if (!dominateMap.containsKey(source)) {
+            // we have no entry yet, compute both values
+            dominates = binding_irdom.block_dominates(source, target);
+
+            HashMap<Pointer, Integer> cachedResults = new HashMap<>();
+            cachedResults.put(target, dominates);
+            dominateMap.put(source, cachedResults);
+
+        } else {
+            Map<Pointer, Integer> cachedResults = dominateMap.get(source);
+            dominates = cachedResults.computeIfAbsent(target, p -> binding_irdom.block_dominates(source, target));
+        }
+        return dominates != 0;
+    }
+
+    private boolean dominatesUses(Node newNode, Node orig) {
+        for (BackEdges.Edge edge : BackEdges.getOuts(orig)) {
+            Node n = edge.node;
+            if (!dominates(newNode.getBlock().ptr, n.getBlock().ptr)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private final class CSEVisitor extends NodeVisitor.Default {
+
+        private final Map<TargetValue, Node> constCache = new HashMap<>();
+        private final Map<Entity, Node> addressCache = new HashMap<>();
+        private final Map<NodePreds, Node> nodeCache = new HashMap<>();
+
+        @RequiredArgsConstructor
+        private class NodePreds {
+            private final Node[] preds;
+            private final binding_irnode.ir_opcode opcode;
+            private final Mode mode;
+
+            @Override
+            public boolean equals(Object o) {
+                if (this == o) return true;
+                if (o == null || getClass() != o.getClass()) return false;
+                NodePreds nodePreds = (NodePreds) o;
+                return Arrays.equals(preds, nodePreds.preds) && opcode == nodePreds.opcode && mode.equals(nodePreds.mode);
+            }
+
+            @Override
+            public int hashCode() {
+                int result = Objects.hash(opcode, mode);
+                result = 31 * result + Arrays.hashCode(preds);
+                return result;
+            }
+        }
+
+        @Override
+        public void defaultVisit(Node node) {
+            // change nothing
+        }
+
+        private void visitPreds(Node node, Node[] preds, binding_irnode.ir_opcode opcode) {
+            NodePreds np = new NodePreds(preds, opcode, node.getMode());
+            if (nodeCache.containsKey(np)) {
+                Node n = nodeCache.get(np);
+                if (dominates(n.getBlock().ptr, node.getBlock().ptr)) {
+                    nodeValues.put(node, n);
+                } else {
+                    // replace and update cache
+                    nodeValues.put(n, node);
+                    nodeCache.put(np, node);
+                }
+            } else {
+                nodeCache.put(np, node);
+            }
+        }
+
+        private Node[] iterableToArray(Iterable<Node> iterable) {
```

Comment:
Please add this as `toArray` to `io.CommonUtil` as a generic function similar to `CommonUtil.toList` (I would suggest using a Stream).